pkg/cmd/admin/migrate/images/imagerefs.go

```diff
@@ -0,0 +1,475 @@
+* buildconfigs         * replicationcontrollers
+* deploymentconfigs    * pods
+* secrets (docker)
+
+Only images, imagestreams, and secrets are updated by default. Updating images and image
+streams requires administrative privileges.`
+
+	internalMigrateImagesExample = `  # Perform a dry-run of migrating all "docker.io" references to "myregistry.com"
+  %[1]s docker.io/*=myregistry.com/*
+
+  # To actually perform the migration, the confirm flag must be appended
+  %[1]s docker.io/*=myregistry.com/* --confirm
+
+  # To see more details of what will be migrated, use the loglevel and output flags
+  %[1]s docker.io/*=myregistry.com/* --loglevel=2 -o yaml
+
+  # Migrate from a service IP to an internal service DNS name
+  %[1]s 172.30.1.54/*=registry.openshift.svc.cluster.local/*
+
+  # Migrate from a service IP to an internal service DNS name for all deployment configs and builds
+  %[1]s 172.30.1.54/*=registry.openshift.svc.cluster.local/* --include=buildconfigs,deploymentconfigs`
+)
+
+type MigrateImageReferenceOptions struct {
+	migrate.ResourceOptions
+
+	Client          client.Interface
+	Mappings        ImageReferenceMappings
+	UpdatePodSpecFn func(obj runtime.Object, fn func(*kapi.PodSpec) error) (bool, error)
+}
+
+// NewCmdMigrateImageReferences implements a MigrateImages command
+func NewCmdMigrateImageReferences(name, fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &MigrateImageReferenceOptions{
+		ResourceOptions: migrate.ResourceOptions{
+			In:      in,
+			Out:     out,
+			ErrOut:  errout,
+			Include: []string{"imagestream", "image", "secrets"},
+		},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s REGISTRY/NAME=REGISTRY/NAME [...]", name),
+		Short:   "Update embedded Docker image references",
+		Long:    internalMigrateImagesLong,
+		Example: fmt.Sprintf(internalMigrateImagesExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Run())
+		},
+	}
+	options.ResourceOptions.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateImageReferenceOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	var remainingArgs []string
+	for _, s := range args {
+		if !strings.Contains(s, "=") {
+			remainingArgs = append(remainingArgs, s)
+			continue
+		}
+		mapping, err := ParseMapping(s)
+		if err != nil {
+			return err
+		}
+		o.Mappings = append(o.Mappings, mapping)
+	}
+
+	o.UpdatePodSpecFn = f.UpdatePodSpecForObject
+
+	if len(remainingArgs) > 0 {
+		return fmt.Errorf("all arguments must be valid FROM=TO mappings")
+	}
+
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c); err != nil {
+		return err
+	}
+
+	osclient, _, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Client = osclient
+
+	return nil
+}
+
+func (o *MigrateImageReferenceOptions) Validate() error {
```

Comment:
The reasoning was, only `Complete` should modify the necessary options struct. The other two should only act upon it, without mutating it in any way. I've spoke with @kargakis about it.